docs/how-tos/how_to_modify_valid_exchanges_in_returns.mdx

```diff
@@ -0,0 +1,68 @@
+---
+sidebar_position: 1
+---
+
+# How to modify valid exchanges in returns
+
+When customers perform a [devolution][returns], they can be presented with
+options to exchange the returned item. We'll see how we can modify which
+variants are prompted as eligible.
+
+To begin with, you can go through the checkout process in the storefront and
+create a completed order. As we need to perform a return, we must capture its
+payments in the system and mark the shipments as delivered. You can do it in
+the backend or through the console:
+
+```ruby
+order = Spree::Order.find_by_number('R723438584')
+order.payments.map(&:capture!)
+order.shipments.map(&:ship!)
+```
+
+If you go now to the backend and try to create a new RMA for that order, you'll
+see that only the same product is shown to the user as a valid exchange. That's
+because, by default,
+[`Spree::ReturnItem::ExchangeVariantEligibility::SameProduct`][same-product] is
+[used as the engine][configure-exchange-variant-engine].
+
+Let's create another engine to allow picking from any variant with shared
+taxons (for instance, exchange a t-shirt with any other t-shirt, but not with a
+teapot). We only need to follow the same API used by the built-in engines.
+Don't forget to ensure that the variant is also in stock:
+
+```ruby title="app/services/amazing_store/return_item/exchange_variant_eligibility/shared_taxon.rb"
+# frozen_string_literal: true
+
+module AmazingStore
+  module ReturnItem
+    module ExchangeVariantEligibility
+      module SameTaxon
+        def self.eligible_variants(variant, stock_locations: nil)
+          Spree::Variant.
+            joins(product: :taxons).
+            where(taxons: { id: variant.product.taxons.ids }, is_master: variant.is_master?).
```

Comment:
This assumes that there are no variants available. I think it's ok for the example, maybe it's worth mentioning this gotcha somewhere? (I'm not super sure of the suggestion here, WDYT?)

Reply:
I'm not sure I follow. Are you talking about the `is_master` condition? To be honest, I didn't fully get why it's needed. I saw it's [used in the default engine](https://github.com/solidusio/solidus/blob/008192cd82dc9e8c270b47782624172ac02b3552/core/app/models/spree/return_item/exchange_variant_eligibility/same_product.rb#L8), and that without it the select contained labels built from options name and values but not the product name.

Reply:
Ah, I just got why it's like that in the engine you linked: if the product has variants, all its purchasable variants will have `is_master?` to `false` and this will let exchange that item with any other variant of that product. If it only has a master variant it will be `true` and it will be exchangeable only with the same master variant (which represent the product in this case). 

For your new example, I think this doesn't work, because there could be products in the same taxon with a different value for `is_master?` than the product we want to exchange. The correct logic would be to pick master variants for products without variants and all the variants except master for products with variants. Does it make sense? Maybe this will make the example more complex, but I'm not sure yet.

Reply:
Oh, yeah, now I understand, and you're right. That would need a very complex query defeating the simplicity we want in the example. I've updated it to leverage the built-in engine and restrict it to variants lighter than the original one.

Reply:
Smart! That's perfect now @waiting-for-dev!